### PR TITLE
Add JWT auth for the api

### PR DESCRIPTION
**Pull request by egnwd**
Serve the user's profile, name in this case, from the api for front end consumption. API endpoints need JWT authentication in order to access the resources.

JWT tokens can be acquired by going to the `authEndpoint + "/jwt/token"`, providing you are an authenticated user.

### CODE REVIEW

**Comment by egnwd** [comment]
ping @plietar

File: auth/jwt.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	"net/http"
+	"strings"
+	"time"
+
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+const (
+	secret = "super secret"
+	issuer = "sponsor-portal.docsoc.co.uk"
```
Comment (plietar) [inline auth/jwt.go]: Both of these should be part of the config

File: auth/jwt.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	json, err := token.SignedString(getSecret())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.Write([]byte(json))
+}
+
+func getExpiry() int64 {
+	return time.Now().Add(time.Hour * 3).Unix()
```
Comment (plietar) [inline auth/jwt.go]: `time.Hour * 3` should be a constant

File: client/src/students.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    Accept: 'application/json',
+    'Content-Type': 'application/json',
+  },
+  credentials: 'same-origin',
+};
+
+const fetchWithConfig = (url, configOverrides) => {
+  const config = Object.assign({}, defaultConfig, configOverrides);
+
+  return fetch(url, config);
```
Comment (plietar) [inline client/src/students.js]: `fetch` is a pretty new API, Safari has only added it a couple of months ago.
Can we use some polyfill for that ? https://github.com/github/fetch

File: client/src/students.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+const getJWTHeader = async (hdr) => {
+  const token = await getToken();
+
+  const headers = hdr || new Headers();
+  headers.append('Authorization', `Bearer ${token}`);
+  return headers;
+}
+
+const getToken = async () => {
+  const response = await fetch("/students/auth/jwt/token", {
```
Comment (plietar) [inline client/src/students.js]: Can't you use `fetchWithConfig` here ?

File: client/src/students.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    credentials: 'same-origin',
+  });
+
+  return response.text();
+}
+
+const fetchUser = async () => {
+  const headers = await getJWTHeader();
```
Comment (plietar) [inline client/src/students.js]: Is this going to fetch a token on every request ?
The token should be cached on the browser (localstorage I guess)

File: model/user.go [hunk 6]
```diff
@@ -9,12 +9,12 @@ import (
 
 type User struct {
 	Id   int64
```
Comment (plietar) [inline model/user.go]: Shouldn't there be a `json:"id"` here ?

File: student/routes.go [hunk 7]
```diff
@@ -1,32 +1,43 @@
...
 	router := mux.NewRouter()
 
 	// auth
-	router.PathPrefix("/auth/").Handler(http.StripPrefix("/auth", service.Auth.Handler()))
+	router.PathPrefix("/auth/").Handler(http.StripPrefix("/auth", s.Auth.Handler()))
 
-	router.Handle("/", service.Auth.RequireAuth(indexHandler(service)))
+	router.Handle("/", s.Auth.RequireAuth(indexHandler(s)))
+
+	api := mux.NewRouter()
```
Comment (plietar) [inline student/routes.go]: Make that into a `getAPIRoutes() http.Handler` function, as it is likely to grow

File: student/routes.go [hunk 8]
```diff
@@ -1,32 +1,43 @@
...
-		user, err := s.UserReader.GetById(id)
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-		}
-		fmt.Fprintf(w, "Hello, %s", user.Name)
+		http.ServeFile(w, r, s.staticFiles+"/students.html")
 	})
 }
+
+func (s *Service) getUserInformation(w http.ResponseWriter, r *http.Request) {
```
Comment (plietar) [inline student/routes.go]: This isn't the right file for this

File: client/src/students.js [hunk 9]
```diff
@@ -26,13 +13,13 @@ const getJWTHeader = async (hdr) => {
   return headers;
 }
 
-const getToken = async () => {
-  const response = await fetch("/students/auth/jwt/token", {
+const getToken = memoize(async () => {
+  const response = await fetchWithConfig("/students/auth/jwt/token", {
     credentials: 'same-origin',
```
Comment (plietar) [inline client/src/students.js]: No need to specify this as it is set by `fetchWithConfig`

File: .env [hunk 10]
```diff
@@ -15,3 +15,6 @@ OAUTH_ISSUER=https://accounts.google.com
 # The following are for the local development only, other domains will not work.
 OAUTH_CLIENT_ID=190477914499-dl3446c9h3t5k5dsidl9t1n1hnekhk3m.apps.googleusercontent.com
 OAUTH_CLIENT_SECRET=lWpdWF2JDHOBD7n4Q1WfyY9g
+
+JWT_SECRET=ecfad4e6356f5d2756a6dce4ce1c7edb
+JWT_ISSUER=sponsor-portal.docsoc.co.u
```
Comment (plietar) [inline .env]: `.uk`

**Review by plietar — APPROVED** [review]
LGTM, @egnwd you can merge this if you don't have anything to add

**Pull request merged**